### PR TITLE
feat(models): deployments_plugin backend (AIRCORE-759 PR1)

### PR DESCRIPTION
**Pull request by tylersbray**
## Summary

- Adds a selectable models `deployments_plugin` `ServiceBackend` that compiles NIM/vLLM/generic deployments into deployments-plugin entities: `Volume` + OnFailure puller + Always server with `prerequisite: succeeded`.
- Projects plugin `Deployment` status onto `ModelDeployment` (trust READY; substrate details in `error_details`); missing server after READY → `LOST`.
- Keeps existing `docker` / `nim_operator` / `none` backends for dual-path rollout. Enable via `models.controller.backends.deployments_plugin.enabled: true` (and disable the other backend).

Linear: [AIRCORE-759](https://linear.app/nvidia/issue/AIRCORE-759/deployments-plugin-models-adopts-the-plugin-drop-the-k8s-nim-operator)

**This is PR1 of 2.** PR2 (delete backends + drop `k8s-nim-operator`) waits for this to merge.

## Resolved open questions

| # | Question | Decision |
|---|----------|----------|
| 1 | Asymmetric projection | Models sole writer of `ModelDeployment` (`spec`/`status`/`status_message`/`error_details`/`model_provider_id`/`host_url`). Plugin owns `Deployment`/`Volume`/`DeploymentConfig`. Models projects plugin status with Layer-3 enrichment; raw substrate in `error_details`. |
| 2 | In-flight upgrade | No alembic. Stale `READY` → `LOST` for manual recreate. No auto-orphan cleanup of old docker containers / NIMService CRs / PVCs. |
| 3 | Executor selection | `docker_executor` / `k8s_executor` / `default_executor` on the new backend config; map `platform.runtime` → named executor. No per-`ModelDeployment` executor field. |
| 4 | LoRA sidecar | Models-owned. k8s: `init_containers` native-sidecar (`restart_policy: Always`). docker: second container emitted (plugin docker v1 is still single-container — LoRA-on-docker via plugin needs a follow-up). |

## Test plan

- [x] Unit: `pytest services/core/models/tests/unit/controllers/backends/deployments_plugin/ -v` (17 tests)
- [x] Registry/config unit tests still green
- [x] `ty check` on new package; ruff clean
- [ ] Enable `deployments_plugin` against a local platform with deployments executors and deploy a vLLM (and NIM) model end-to-end
- [ ] Confirm READY → LOST when plugin server Deployment is deleted under a READY ModelDeployment
- [ ] Confirm delete ordering leaves ModelDeployment until plugin Deployments are gone

## Enablement sketch

```yaml
models:
  controller:
    backends:
      docker:
        enabled: false
      deployments_plugin:
        enabled: true
        docker_executor: local-docker   # or k8s_executor: local-k8s
```

## Summary by CodeRabbit

* **New Features**
  * Added an optional Deployments Plugin backend for creating, monitoring, and removing model deployments.
  * Supports Docker and Kubernetes execution, vLLM and NIM serving, optional weight pulling, and LoRA integrations.
* **Documentation**
  * Expanded the configuration reference with deployments-plugin settings (timeouts, images/tags, storage defaults, and LoRA controls), including an enablement switch.
* **Bug Fixes**
  * Added fail-fast validation for unsupported Docker+LoRA combinations.
  * Improved drift recovery/teardown behavior and pending-timeout escalation.
* **Tests**
  * Added unit coverage for backend execution, compilation, naming, status mapping, and edge cases.

### CODE REVIEW

File: services/core/models/src/nmp/core/models/controllers/backends/deployments_plugin/__init__.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+"""
+
+from typing import Any
+
+from .config import DeploymentsPluginBackendConfigModel as DeploymentsPluginBackendConfigModel
+
+__all__ = ["DeploymentsPluginBackendConfigModel", "DeploymentsPluginServiceBackend"]
+
+
+def __getattr__(name: str) -> Any:
```
Comment (mckornfield) [inline services/core/models/src/nmp/core/models/controllers/backends/deployments_plugin/__init__.py]: lol this is sorta weird

File: services/core/models/src/nmp/core/models/config.py [hunk 2]
```diff
@@ -163,12 +164,13 @@ class ParallelismConfig(BaseModel):
 # Backend and controller configuration
 # -----------------------------------------------------------------------------
 
-BackendName = Literal["docker", "nim_operator"]
+BackendName = Literal["docker", "nim_operator", "deployments_plugin", "none"]
```
Comment (mckornfield) [inline services/core/models/src/nmp/core/models/config.py:167]: what does the `none` backend do, break or noop?

File: services/core/models/src/nmp/core/models/controllers/backends/registry.py [hunk 3]
```diff
@@ -36,19 +42,45 @@ class NoneBackendConfigModel(BaseModel):
...
+    name: BackendName, available_backends: Dict[BackendName, type[ServiceBackend]]
+) -> type[ServiceBackend]:
+    """Return the backend class for ``name``, importing optional backends lazily."""
+    if name in available_backends:
+        return available_backends[name]
+    if name == "deployments_plugin":
+        from nmp.core.models.controllers.backends.deployments_plugin import DeploymentsPluginServiceBackend
+
+        return DeploymentsPluginServiceBackend
+    available = ", ".join(sorted({*available_backends, *_LAZY_BACKEND_NAMES}))
```
Comment (mckornfield) [inline services/core/models/src/nmp/core/models/controllers/backends/registry.py:96]: this is sorta weird. it also might not be available if it's optional yeah? should we give guidance in that case?

File: services/core/models/src/nmp/core/models/controllers/backends/deployments_plugin/resolve.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+from nmp.core.models.controllers.context import ModelContext
+
+
+@dataclass(frozen=True)
+class ResolvedPluginDeployment:
+    """All API-object data required to compile plugin entities."""
+
+    deployment: ModelDeployment
+    config: ModelDeploymentConfig
+    model_entity: Optional[ModelEntity]
```
Comment (mckornfield) [inline services/core/models/src/nmp/core/models/controllers/backends/deployments_plugin/resolve.py]: should it just be
```suggestion
    model_entity: ModelEntity | None
```

File: services/core/models/src/nmp/core/models/controllers/backends/deployments_plugin/resolve.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    view = deployment_config_view(ctx.model_deployment_config)
+    namespace, name, revision = resolve_model_source(ctx.model_entity, view)
+    platform_config = get_platform_config()
+    files_url = platform_config.service_discovery.get("files") or platform_config.base_url
+    return ResolvedPluginDeployment(
+        deployment=ctx.model_deployment,
+        config=ctx.model_deployment_config,
+        model_entity=ctx.model_entity,
+        view=view,
+        weights_type=get_model_weights_type(
```
Comment (mckornfield) [inline services/core/models/src/nmp/core/models/controllers/backends/deployments_plugin/resolve.py:63]: nit- would prefer extracting this and the files_hf_url to vars for readability

File: services/core/models/src/nmp/core/models/controllers/backends/deployments_plugin/executor.py [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+"""Executor selection for deployments-plugin model entities."""
+
+from nmp.common.config import Runtime
+from nmp.core.models.controllers.backends.deployments_plugin.config import DeploymentsPluginConfig
+
+
+def executor_for_runtime(config: DeploymentsPluginConfig, runtime: Runtime) -> str | None:
+    """Resolve a configured executor for the platform runtime."""
+    if runtime == Runtime.DOCKER:
+        return config.docker_executor or config.default_executor
```
Comment (mckornfield) [inline services/core/models/src/nmp/core/models/controllers/backends/deployments_plugin/executor.py:21]: what scenario would you choose a runtime, but then the executor wouldn't be configured?

File: services/core/models/src/nmp/core/models/controllers/backends/deployments_plugin/compiler.py [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+            workspace=resolved.deployment.workspace,
+            size=_SCRATCH_VOLUME_SIZE,
+            backendConfig=VolumeBackendConfig(k8s=K8sVolumeConfig(storageClass=config.default_storage_class)),
+        )
+        mounts.append(VolumeMount(name=names.scratch, mountPath=_SCRATCH_MOUNT))
+        # Ensure the LoRA cache dir exists before the server/sidecar start.
+        init_containers.append(
+            Container(
+                name="lora-cache-init",
+                image="busybox:1.36",
```
Comment (mckornfield) [inline services/core/models/src/nmp/core/models/controllers/backends/deployments_plugin/compiler.py]: nit- do we want to hard code an image like this? what if there are restrictions on hitting docker.io? I assume we should make it configurable

File: services/core/models/src/nmp/core/models/controllers/backends/deployments_plugin/backend.py [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+
+    async def _rollback_create(self, ctx: ModelContext) -> None:
+        """Best-effort controlled teardown after a partial create failure."""
+        if ctx.model_deployment is None:
+            return
+        try:
+            await self.delete_model_deployment(ctx.model_deployment.workspace, ctx.model_deployment.name)
+        except Exception:
+            logger.warning(
+                "Failed to roll back deployments-plugin substrate for %s/%s after a create failure; "
```
Comment (mckornfield) [inline services/core/models/src/nmp/core/models/controllers/backends/deployments_plugin/backend.py]: nit- do we use structured logging here?

**Review by mckornfield — APPROVED** [review]

File: services/core/models/src/nmp/core/models/controllers/backends/deployments_plugin/__init__.py (hunk 1 above)
Comment (tylersbray) [reply services/core/models/src/nmp/core/models/controllers/backends/deployments_plugin/__init__.py]: Removed __getattr__; package only exports config. Registry lazy-imports backend directly.

File: services/core/models/src/nmp/core/models/config.py (hunk 2 above)
Comment (tylersbray) [reply services/core/models/src/nmp/core/models/config.py:167]: Documented NoneBackendConfigModel as intentional no-op for platform.runtime: none (CRUD raises, status returns UNKNOWN). Added comment on BACKEND_CONFIG_MODELS.

File: services/core/models/src/nmp/core/models/controllers/backends/registry.py (hunk 3 above)
Comment (tylersbray) [reply services/core/models/src/nmp/core/models/controllers/backends/registry.py:96]: _resolve_backend_class catches ImportError with install guidance. New unit test verifies the message.

File: services/core/models/src/nmp/core/models/controllers/backends/deployments_plugin/resolve.py (hunk 4 above)
Comment (tylersbray) [reply services/core/models/src/nmp/core/models/controllers/backends/deployments_plugin/resolve.py]: Updated ResolvedPluginDeployment.

File: services/core/models/src/nmp/core/models/controllers/backends/deployments_plugin/resolve.py (hunk 5 above)
Comment (tylersbray) [reply services/core/models/src/nmp/core/models/controllers/backends/deployments_plugin/resolve.py:63]: files_service_url and files_hf_url are now named locals in resolve_plugin_deployment.

File: services/core/models/src/nmp/core/models/controllers/backends/deployments_plugin/executor.py (hunk 6 above)
Comment (tylersbray) [reply services/core/models/src/nmp/core/models/controllers/backends/deployments_plugin/executor.py:21]: Expanded executor_for_runtime docstring (misconfiguration when executor names are omitted).

File: services/core/models/src/nmp/core/models/controllers/backends/deployments_plugin/compiler.py (hunk 7 above)
Comment (tylersbray) [reply services/core/models/src/nmp/core/models/controllers/backends/deployments_plugin/compiler.py]: Added configurable busybox_image / busybox_image_tag (default docker.io/library/busybox:latest).

File: services/core/models/src/nmp/core/models/controllers/backends/deployments_plugin/backend.py (hunk 8 above)
Comment (tylersbray) [reply services/core/models/src/nmp/core/models/controllers/backends/deployments_plugin/backend.py]: Rollback warning now uses extra={workspace, deployment_name}.
